{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a74baf28-c2d5-43b8-98a6-1b500ab11568",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import shap\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from pycaret.classification import *\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import cross_val_score, RepeatedKFold, StratifiedKFold, train_test_split, GridSearchCV, RandomizedSearchCV\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import f_classif\n",
    "from tsfresh import extract_features, select_features\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import mannwhitneyu,ttest_ind\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import gzip\n",
    "import io\n",
    "from matplotlib.patches import Circle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bae83a1d-2b8b-4e07-9c0c-c4e9929d4c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "balanced = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d0b0473-0bb6-4afd-9bdb-e3ac1cdc8453",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_matrix = pd.read_csv('FinalData/SFARI_LDfirst_mutation_matrix_normalised_rnd_noXY.txt', sep=',')\n",
    "gene_matrix_non_norm = pd.read_csv('FinalData/SFARI_LDfirst_mutation_matrix_noXY.txt', sep=',')\n",
    "snp_matrix = pd.read_csv('FinalData/small_LDpruned_allSNPs5e-5_training_inc_noID.raw', sep=' ')\n",
    "prs_matrix = pd.read_csv('FinalData/all_CG_PRS_SNPs.raw', sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eda328f-2810-45c4-83f0-7fc577f43724",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Split gene data into 56:24:20\n",
    "y = gene_matrix['PHENOTYPE']\n",
    "X = gene_matrix.drop(columns=['PHENOTYPE', 'SEX', 'IID'])\n",
    "val = []\n",
    "for i in range(len(y)):\n",
    "    if y[i] != 1 and y[i] != 2:\n",
    "        val.append(i)\n",
    "y = y.drop(val)\n",
    "X = X.drop(val)\n",
    "gene_matrix = gene_matrix.drop(val)\n",
    "y = [pheno-1 for pheno in y]\n",
    "X = X.fillna(0)\n",
    "gene_matrix = gene_matrix.fillna(0)\n",
    "\n",
    "X_trainG, X_testG, y_trainG, y_testG = train_test_split(X,y,test_size=0.2, random_state=10)\n",
    "X_train2G, X_test2G, y_train2G, y_test2G = train_test_split(X_trainG,y_trainG,test_size=0.3, random_state=10)\n",
    "trainingG = X_train2G.copy()\n",
    "trainingG['PHENOTYPE'] = y_train2G\n",
    "testG = X_test2G.copy()\n",
    "testG['PHENOTYPE'] = y_test2G\n",
    "\n",
    "unseen_testG = X_testG.copy()\n",
    "unseen_testG['PHENOTYPE'] = y_testG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad2b825f-62ef-48c9-b907-f8ffca03f9b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subsample the males, if balanced\n",
    "if balanced:\n",
    "    trainingG_aut = trainingG[trainingG['PHENOTYPE'] == 1]\n",
    "    trainingG_neuro = trainingG[trainingG['PHENOTYPE'] != 1]  # Keep data with other phenotypes\n",
    "\n",
    "    males_aut = trainingG_aut[gene_matrix.loc[trainingG_aut.index, 'SEX'] == 1]\n",
    "    females_aut = trainingG_aut[gene_matrix.loc[trainingG_aut.index, 'SEX'] == 2]\n",
    "    n_males = len(males_aut)\n",
    "    males_aut_balanced = males_aut.sample(n=len(females_aut), random_state=10)\n",
    "    balanced_aut_data = pd.concat([males_aut_balanced, females_aut])\n",
    "    balanced_trainingG = pd.concat([balanced_aut_data, trainingG_neuro])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1fbdf0-7a62-49b6-8d63-3bd3fd24f81a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace '.' names from Fisher SNPs with the position\n",
    "traw = pd.read_csv('FinalData/small_LDpruned_allSNPs5e-5_training_inc_noID.traw', sep = '\\t')\n",
    "\n",
    "mapping = {}\n",
    "for colno in range(len(snp_matrix.columns)):\n",
    "    base_name = snp_matrix.columns[colno].split('_')[0]  \n",
    "    if base_name == '.':\n",
    "        new_name = f\"{traw['CHR'].values[colno-6]}_{traw['POS'].values[colno-6]}_{traw['COUNTED'].values[colno-6]}\"\n",
    "    else:\n",
    "        new_name = snp_matrix.columns[colno]\n",
    "    \n",
    "    mapping[snp_matrix.columns[colno]] = new_name\n",
    "snp_matrix.rename(columns=mapping, inplace=True)\n",
    "\n",
    "# Split SNP data into 56:24:20\n",
    "snp_y = snp_matrix['PHENOTYPE']\n",
    "snp_X = snp_matrix.drop(columns=['FID', 'IID', 'PAT', 'MAT', 'PHENOTYPE', 'SEX'])\n",
    "val = []\n",
    "for i in range(len(snp_y)):\n",
    "    if snp_y[i] != 1 and snp_y[i] != 2:\n",
    "        val.append(i)\n",
    "snp_y = snp_y.drop(val)\n",
    "snp_X = snp_X.drop(val)\n",
    "snp_matrix = snp_matrix.drop(val)\n",
    "snp_y = [pheno-1 for pheno in snp_y]\n",
    "snp_X = snp_X.fillna(0)\n",
    "\n",
    "X_trainS, X_testS, y_trainS, y_testS = train_test_split(snp_X,snp_y,test_size=0.2, random_state=10)\n",
    "X_train2S, X_test2S, y_train2S, y_test2S = train_test_split(X_trainS,y_trainS, test_size=0.3, random_state=10)\n",
    "trainingS = X_train2S.copy()\n",
    "trainingS['PHENOTYPE'] = y_train2S\n",
    "testS = X_test2S.copy()\n",
    "testS['PHENOTYPE'] = y_test2S\n",
    "\n",
    "unseen_testS = X_testS.copy()\n",
    "unseen_testS['PHENOTYPE'] = y_testS\n",
    "\n",
    "if balanced:\n",
    "    balanced_trainingS = trainingS.loc[balanced_trainingG.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd90f10-215c-4af7-8965-7f6c2b10e53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split PRS data into 56:24:20\n",
    "prs_y = prs_matrix['PHENOTYPE']\n",
    "prs_X = prs_matrix.drop(columns=['FID', 'IID', 'PAT', 'MAT', 'PHENOTYPE'])\n",
    "val = []\n",
    "for i in range(len(prs_y)):\n",
    "    if prs_y[i] != 1 and prs_y[i] != 2:\n",
    "        val.append(i)\n",
    "prs_y = prs_y.drop(val)\n",
    "prs_X = prs_X.drop(val)\n",
    "prs_matrix = prs_matrix.drop(val)\n",
    "prs_y = [pheno-1 for pheno in prs_y]\n",
    "prs_X = prs_X.fillna(0)\n",
    "\n",
    "X_trainP, X_testP, y_trainP, y_testP = train_test_split(prs_X,prs_y,test_size=0.2, random_state=10)\n",
    "X_train2P, X_test2P, y_train2P, y_test2P = train_test_split(X_trainP,y_trainP, test_size=0.3, random_state=10)\n",
    "trainingP = X_train2P.copy()\n",
    "trainingP['PHENOTYPE'] = y_train2P\n",
    "testP = X_test2P.copy()\n",
    "testP['PHENOTYPE'] = y_test2P\n",
    "\n",
    "unseen_testP = X_testP.copy()\n",
    "unseen_testP['PHENOTYPE'] = y_testP\n",
    "\n",
    "if balanced:\n",
    "    balanced_trainingP = trainingP.loc[balanced_trainingG.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ee1ae40-61e8-4e96-a605-0605b6e1902b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create rf model from gene data and get predictive scores\n",
    "gene = ClassificationExperiment()\n",
    "if balanced:\n",
    "    gene.setup(balanced_trainingG, target = 'PHENOTYPE', session_id=10, test_data = testG)\n",
    "else:\n",
    "    gene.setup(trainingG, target = 'PHENOTYPE', session_id=10, test_data = testG)\n",
    "rf = gene.create_model('rf')\n",
    "tuned_rf = gene.tune_model(rf, optimize = 'F1')\n",
    "pred_gene = gene.predict_model(tuned_rf)\n",
    "pred_test_gene = gene.predict_model(tuned_rf, data = unseen_testG)\n",
    "pred_train_gene = gene.predict_model(tuned_rf, data=trainingG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed66225b-3197-4468-b2a7-fb85c3fd93b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select best features from Fisher SNP dataset using MWU on training data\n",
    "X_train_mwu = select_features(X_train2S, np.array(y_train2S),test_for_binary_target_real_feature='mann', test_for_real_target_binary_feature='mann', fdr_level=0.05)\n",
    "X_test_mwu = X_test2S[X_test2S.columns.intersection(X_train_mwu.columns)]\n",
    "X_test_mwu = X_test_mwu.reindex(sorted(X_test_mwu.columns), axis=1)\n",
    "X_train_mwu = X_train_mwu.reindex(sorted(X_train_mwu.columns), axis=1)\n",
    "\n",
    "trainingS = X_train_mwu\n",
    "trainingS['PHENOTYPE'] = y_train2S\n",
    "testS = X_test_mwu\n",
    "testS['PHENOTYPE'] = y_test2S\n",
    "\n",
    "if balanced:\n",
    "    balanced_trainingS = balanced_trainingS[trainingS.columns]\n",
    "\n",
    "# create rf model from fisher SNP dataset and get predictive scores\n",
    "snp = ClassificationExperiment()\n",
    "if balanced:\n",
    "    snp.setup(balanced_trainingS, target = 'PHENOTYPE', session_id=10, test_data = testS)\n",
    "else:\n",
    "    snp.setup(trainingS, target = 'PHENOTYPE', session_id=10, test_data = testS)\n",
    "rfS = snp.create_model('rf')\n",
    "tuned_rfS = snp.tune_model(rfS, optimize = 'F1')\n",
    "pred_snp = snp.predict_model(tuned_rfS)\n",
    "pred_test_snp = snp.predict_model(tuned_rfS, data=unseen_testS)\n",
    "pred_train_snp = snp.predict_model(tuned_rfS, data=trainingS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56605d3d-5b28-4f7f-a504-95d7954b67a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create rf model from PRS dataset and get predictive scores\n",
    "prs = ClassificationExperiment()\n",
    "if balanced:\n",
    "    prs.setup(balanced_trainingP, target = 'PHENOTYPE', session_id=10, test_data = testP)\n",
    "else:\n",
    "    prs.setup(trainingP, target = 'PHENOTYPE', session_id=10, test_data = testP)\n",
    "rfP = prs.create_model('rf')\n",
    "tuned_rfP = prs.tune_model(rfP, optimize = 'F1')\n",
    "pred_prs = prs.predict_model(tuned_rfP)\n",
    "pred_test_prs = prs.predict_model(tuned_rfP, data=unseen_testP)\n",
    "pred_train_prs = prs.predict_model(tuned_rfP, data=trainingP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99cfad44-104e-4d9f-bd16-91f4c639c264",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create datasets for ensemble model made up of the predictive scores from the previous three rf and biological sex\n",
    "#dataset that will be used for ensemble training\n",
    "comb_pred = pd.DataFrame()\n",
    "pred_gene = pred_gene.reset_index(drop=True)\n",
    "comb_pred['gene'] = pred_gene['prediction_score']\n",
    "gene_predictions = pred_gene['prediction_label']\n",
    "for i in range(len(gene_predictions)):\n",
    "  if gene_predictions[i] == 0:\n",
    "    comb_pred['gene'][i] = comb_pred['gene'][i]*-1 # if predicted neurotypical, make score negative\n",
    "\n",
    "pred_snp = pred_snp.reset_index(drop=True)\n",
    "comb_pred['snp'] = pred_snp['prediction_score']\n",
    "snp_predictions = pred_snp['prediction_label']\n",
    "for i in range(len(snp_predictions)):\n",
    "  if snp_predictions[i] == 0:\n",
    "    comb_pred['snp'][i] = comb_pred['snp'][i]*-1\n",
    "\n",
    "pred_prs = pred_prs.reset_index(drop=True)\n",
    "comb_pred['prs'] = pred_prs['prediction_score']\n",
    "prs_predictions = pred_prs['prediction_label']\n",
    "for i in range(len(prs_predictions)):\n",
    "  if prs_predictions[i] == 0:\n",
    "    comb_pred['prs'][i] = comb_pred['prs'][i]*-1\n",
    "\n",
    "comb_pred['SEX'] = pred_prs['SEX']\n",
    "comb_pred['PHENOTYPE'] = pred_snp['PHENOTYPE']\n",
    "#dataset for unseen test\n",
    "comb_test = pd.DataFrame()\n",
    "pred_test_gene = pred_test_gene.reset_index(drop=True)\n",
    "comb_test['gene'] = pred_test_gene['prediction_score']\n",
    "gene_predictionsT = pred_test_gene['prediction_label']\n",
    "for i in range(len(gene_predictionsT)):\n",
    "  if gene_predictionsT[i] == 0:\n",
    "   comb_test['gene'][i] = comb_test['gene'][i]*-1\n",
    "\n",
    "pred_test_snp = pred_test_snp.reset_index(drop=True)\n",
    "comb_test['snp'] = pred_test_snp['prediction_score']\n",
    "snp_predictionsT = pred_test_snp['prediction_label']\n",
    "for i in range(len(snp_predictionsT)):\n",
    "  if snp_predictionsT[i] == 0:\n",
    "    comb_test['snp'][i] = comb_test['snp'][i]*-1\n",
    "\n",
    "pred_test_prs = pred_test_prs.reset_index(drop=True)\n",
    "comb_test['prs'] = pred_test_prs['prediction_score']\n",
    "prs_predictionsT = pred_test_prs['prediction_label']\n",
    "for i in range(len(prs_predictionsT)):\n",
    "  if prs_predictionsT[i] == 0:\n",
    "    comb_test['prs'][i] = comb_test['prs'][i]*-1\n",
    "\n",
    "comb_test['SEX'] = pred_test_prs['SEX']\n",
    "comb_test['PHENOTYPE'] = pred_test_snp['PHENOTYPE']\n",
    "\n",
    "#dataset that the og models were trained on\n",
    "comb_train = pd.DataFrame()\n",
    "pred_train_gene = pred_train_gene.reset_index(drop=True)\n",
    "comb_train['gene'] = pred_train_gene['prediction_score']\n",
    "gene_predictions = pred_train_gene['prediction_label']\n",
    "for i in range(len(gene_predictions)):\n",
    "  if gene_predictions[i] == 0:\n",
    "    comb_train['gene'][i] = comb_train['gene'][i]*-1\n",
    "\n",
    "pred_train_snp = pred_train_snp.reset_index(drop=True)\n",
    "comb_train['snp'] = pred_train_snp['prediction_score']\n",
    "snp_predictions = pred_train_snp['prediction_label']\n",
    "for i in range(len(snp_predictions)):\n",
    "  if snp_predictions[i] == 0:\n",
    "    comb_train['snp'][i] = comb_train['snp'][i]*-1\n",
    "\n",
    "pred_train_prs = pred_train_prs.reset_index(drop=True)\n",
    "comb_train['prs'] = pred_train_prs['prediction_score']\n",
    "prs_predictions = pred_train_prs['prediction_label']\n",
    "for i in range(len(prs_predictions)):\n",
    "  if prs_predictions[i] == 0:\n",
    "    comb_train['prs'][i] = comb_train['prs'][i]*-1\n",
    "\n",
    "comb_train['SEX'] = pred_train_prs['SEX']\n",
    "comb_train['PHENOTYPE'] = pred_train_snp['PHENOTYPE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71aa391d-548b-4a0c-b3fa-b56fd77030ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# train and predict using ensemble model\n",
    "comb = ClassificationExperiment()\n",
    "comb.setup(comb_pred, target = 'PHENOTYPE', session_id=10, test_data = comb_test, index=False)\n",
    "rfC = comb.create_model('rf')\n",
    "tuned_rfC = comb.tune_model(rfC, optimize = 'F1')\n",
    "predC = comb.predict_model(tuned_rfC)\n",
    "predC_test = comb.predict_model(tuned_rfC, data=comb_pred)\n",
    "predC_train = comb.predict_model(tuned_rfC, data=comb_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a46b40-3cad-47d1-b279-be1008f91892",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how well does it predict for each sex?\n",
    "predC['SEX'] = snp_matrix['SEX']\n",
    "aut_pred = predC[predC['prediction_label'] == 1]\n",
    "aut_corr = aut_pred[aut_pred['PHENOTYPE']==1]\n",
    "print('Males predicted autistic:\\t',len(aut_pred[aut_pred['SEX'] == 1]))\n",
    "print('What percentage of these are correct:\\t', len(aut_corr[aut_corr['SEX']==1])/len(aut_pred[aut_pred['SEX'] == 1]))\n",
    "print('Females predicted autistic:',len(aut_pred[aut_pred['SEX'] == 2]))\n",
    "print('What percentage of these are correct:\\t', len(aut_corr[aut_corr['SEX']==2])/len(aut_pred[aut_pred['SEX'] == 2]))\n",
    "\n",
    "naut_pred = predC[predC['prediction_label'] == 0]\n",
    "naut_corr = naut_pred[naut_pred['PHENOTYPE']==0]\n",
    "print('Males predicted neurotypical:\\t',len(naut_pred[naut_pred['SEX'] == 1]))\n",
    "print('What percentage of these are correct:\\t', len(naut_corr[naut_corr['SEX']==1])/len(naut_pred[naut_pred['SEX'] == 1]))\n",
    "print('Females predicted neurotypical:',len(naut_pred[naut_pred['SEX'] == 2]))\n",
    "print('What percentage of these are correct:\\t', len(naut_corr[naut_corr['SEX']==2])/len(naut_pred[naut_pred['SEX'] == 2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb4399b-83ba-4c6a-b8bc-9bb24ab19934",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get shap values for overall model\n",
    "save = False\n",
    "explainer = shap.Explainer(tuned_rfC)\n",
    "X_test_comb = comb_test.drop(columns='PHENOTYPE')\n",
    "shap_values = explainer.shap_values(X_test_comb)\n",
    "shap.summary_plot(shap_values[:,:,1], X_test_comb, show=False)\n",
    "ax = plt.gca()\n",
    "ax.set_xlim(-0.28,0.17)\n",
    "if save:\n",
    "    plt.savefig('shap_summary.png', dpi=700)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2a33a7-a9d4-483f-ae79-4565c8bf43c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dependence plot of impact of biological snp on fisher SNP\n",
    "save = False\n",
    "shap.dependence_plot(\"snp\", shap_values[:,:,1], X_test_comb,interaction_index=\"SEX\", show=False)\n",
    "ax = plt.gca()\n",
    "ax.set_xlim(-1,1)\n",
    "ax.set_ylim(-0.28,0.18)\n",
    "if save:\n",
    "    plt.savefig('snp_sex_dependence.png', dpi=700)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a2dca4-df80-4d77-b05e-7b0c33bbe552",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find the locations of all the SNPs in SNP matrix. Since some are '.' need to download the matrix with the locations\n",
    "positions = pd.read_csv('FinalData/plink.assoc1711LD0.5_small.fisher', delim_whitespace=True)\n",
    "chosen_snps = pd.read_csv('FinalData/small_LDpruned_allSNPs5e-5_training_inc_noID.txt', sep = '\\t', header=None)\n",
    "\n",
    "chosen_snps = chosen_snps.drop(columns=[2,3])\n",
    "chosen_snps.columns = ['chrom', 'bp']\n",
    "chosen_snps = pd.merge(chosen_snps, positions, left_on=['chrom', 'bp'], right_on=['CHR', 'BP'], how='left').drop(columns=['chrom', 'bp', 'A1', 'F_A', 'F_U', 'A2', 'P', 'OR'])\n",
    "\n",
    "for rowno in range(len(chosen_snps)):\n",
    "    if chosen_snps['SNP'][rowno] == '.':\n",
    "        chosen_snps['SNP'][rowno] = f\"{chosen_snps['CHR'][rowno]}_{chosen_snps['BP'][rowno]}\"\n",
    "\n",
    "# Get shap scores for SNPs\n",
    "explainerS = shap.Explainer(tuned_rfS)\n",
    "X_test_S = testS.drop(columns=['PHENOTYPE'])\n",
    "shap_valuesS = explainerS.shap_values(X_test_S)\n",
    "mean_shap_feature_values = pd.DataFrame(shap_valuesS[:,:,1], columns=X_test_S.columns).abs().mean(axis=0).sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596fa7f8-8fa5-4319-93d3-f17d6efbaa18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine shap scores with chr_bp df\n",
    "shap_scoresS = pd.DataFrame(mean_shap_feature_values).reset_index().rename(columns={'index':'snp', 0:'shap'})\n",
    "shap_scoresS['snp_clean'] = shap_scoresS['snp'].str.rsplit('_', n=1).str[0]\n",
    "\n",
    "def replace_with_chr_bp(row):\n",
    "    if row.startswith('rs'):\n",
    "        match = chosen_snps[chosen_snps['SNP'] == row]\n",
    "        if not match.empty:\n",
    "            chr_value = match.iloc[0]['CHR']\n",
    "            bp_value = match.iloc[0]['BP']\n",
    "            return f\"{chr_value}_{bp_value}\"\n",
    "    return row\n",
    "shap_scoresS['snp_clean'] = shap_scoresS['snp_clean'].apply(replace_with_chr_bp)\n",
    "\n",
    "# separate into sex and plot the ten most contributory features\n",
    "pos_class = shap_valuesS[:,:,1]\n",
    "shap_df = pd.DataFrame(pos_class, columns=X_test_S.columns)\n",
    "shap_df['sex'] = testP['SEX'].values\n",
    "\n",
    "shap_female = shap_df[shap_df['sex'] == 2].drop('sex', axis=1)\n",
    "shap_male = shap_df[shap_df['sex'] == 1].drop('sex', axis=1) \n",
    "\n",
    "mean_shap_female = shap_female.abs().mean()\n",
    "mean_shap_male = shap_male.abs().mean()\n",
    "\n",
    "shap_comparison = pd.DataFrame({\n",
    "    'female': mean_shap_female,\n",
    "    'male': mean_shap_male\n",
    "})\n",
    "\n",
    "top_n = 10\n",
    "shap_comparison.sort_values(by='male', ascending=False).head(top_n).plot(kind='bar', figsize=(10, 6))\n",
    "plt.title(f'Top {top_n} Feature Importance Comparison (Female vs Male)')\n",
    "plt.ylabel('Mean |SHAP Value|')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4776169-f7d7-4e8c-aef1-fefa376a05b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "save = False\n",
    "# Calculate top 10% contributors for male and female separately. Plot where they are \n",
    "mean_shap_female = mean_shap_female.sort_values(ascending=False)\n",
    "mean_shap_male = mean_shap_male.sort_values(ascending=False)\n",
    "shap_female_top10 = mean_shap_female[:int(len(mean_shap_female)/10)]\n",
    "shap_male_top10 = mean_shap_male[:int(len(mean_shap_male)/10)]\n",
    "# Which are same and different?\n",
    "common_top10 = set(shap_female_top10.index.values.tolist()) & set(shap_male_top10.index.values.tolist())\n",
    "fem_only_top10 = set(shap_female_top10.index.values.tolist()) - set(shap_male_top10.index.values.tolist())\n",
    "mal_only_top10 = set(shap_male_top10.index.values.tolist()) - set(shap_female_top10.index.values.tolist())\n",
    "def top10_pos(top10, typ):\n",
    "    top10 = pd.DataFrame(top10)\n",
    "    top10['Phenotype'] = typ\n",
    "    top10 = top10.rename(columns={0: 'SNP'})\n",
    "    top10['SNP'] = top10['SNP'].str.rsplit('_', n=1).str[0]\n",
    "    top10 = top10.merge(chosen_snps, how='inner', on='SNP')\n",
    "    return(top10)\n",
    "common_top10 = top10_pos(common_top10,'Shared')\n",
    "fem_only_top10 = top10_pos(fem_only_top10,'Female Only')\n",
    "mal_only_top10 = top10_pos(mal_only_top10, 'Male Only')\n",
    "all_top10_shap = pd.concat([fem_only_top10, mal_only_top10, common_top10])\n",
    "all_top10_shap = all_top10_shap.reset_index(drop=True)\n",
    "if save:\n",
    "    all_top10_shap.to_csv('top10percent_snps_from_model.txt', sep = '\\t', index=False)\n",
    "##Phenogram made from this using https://visualization.ritchielab.org/phenograms/create"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b2363a4-d266-4be8-9855-a718d1223533",
   "metadata": {},
   "source": [
    "From here, only done for balanced model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7970ab50-bf65-44ec-8a16-1f2388091228",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for sig difference for rs58741612_A\n",
    "t_stat, p_value = ttest_ind(abs(shap_female['rs58741612_A']), abs(shap_male['rs58741612_A']), equal_var=False)\n",
    "print(f\"T-test p-value: {p_value}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f96f5c3d-6ece-4eff-9ff3-7c8ed0377c72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify SNPs in chr16p11 then check for sig locus\n",
    "snp1 = [col for col in shap_female.columns if col.startswith('rs73545273')][0]\n",
    "snp2 = [col for col in shap_female.columns if col.startswith('16_34145852')][0]\n",
    "snp3 = [col for col in shap_female.columns if col.startswith('16_34120022')][0]\n",
    "\n",
    "fem_sum = abs(shap_female[snp1] + shap_female[snp2] + shap_female[snp3])\n",
    "mal_sum = abs(shap_male[snp1] + shap_male[snp2] + shap_male[snp3])\n",
    "\n",
    "# is there a difference in contribution for males vs females?\n",
    "t_stat, p_value = ttest_ind(fem_sum, mal_sum, equal_var=False)\n",
    "print(f\"T-test p-value: {p_value}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e90936b-80ab-4e77-9a5a-906efda43c01",
   "metadata": {},
   "source": [
    "Analysing gene mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bb56460-6cc1-4d6f-942b-97ed428f00e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sig_genes(genes_aut, genes_neuro):\n",
    "    gene_names = genes_aut.columns\n",
    "    mwu = {}\n",
    "\n",
    "    for gene in gene_names:\n",
    "        stat, p_value = mannwhitneyu(genes_aut[gene], genes_neuro[gene], alternative='two-sided')\n",
    "        mwu[gene] = {\"statistic\": stat, \"p_val\": p_value}\n",
    "\n",
    "    mwu_res = pd.DataFrame(mwu).T\n",
    "\n",
    "    reject, pvals_corrected, _, _ = multipletests(mwu_res['p_val'], method='fdr_bh')\n",
    "\n",
    "    mwu_res[\"p_val_corr\"] = pvals_corrected\n",
    "    mwu_res[\"sig_corr\"] = reject\n",
    "    return(mwu_res)\n",
    "\n",
    "def loeuf_score_calc(aut_genes, neuro_genes):\n",
    "    all_mwu_res = sig_genes(aut_genes, neuro_genes)\n",
    "    signif_genes = all_mwu_res[all_mwu_res['sig_corr']==True]\n",
    "    insignif_genes = all_mwu_res[all_mwu_res['sig_corr']==False]\n",
    "    print(len(signif_genes))\n",
    "\n",
    "    signif_genes = signif_genes.join(loeuf_scores[['oe_lof_upper','chromosome']], how='left')\n",
    "    signif_genes = signif_genes[~signif_genes['oe_lof_upper'].isna()]\n",
    "    insignif_genes = insignif_genes.join(loeuf_scores[['oe_lof_upper','chromosome']], how='left')\n",
    "    insignif_genes = insignif_genes[~insignif_genes['oe_lof_upper'].isna()]\n",
    "    return(signif_genes, insignif_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5435ba30-4317-4dd7-93f7-37598eb1cec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOEUF scores downloaded from gnomad\n",
    "file_path = 'FinalData/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz'\n",
    "with gzip.open(file_path, 'rt') as f:  \n",
    "        loeuf_scores = pd.read_csv(f, sep='\\t') \n",
    "loeuf_scores = loeuf_scores.set_index(['gene'])            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a456c2-8741-4d90-80ec-bd7400d8ef2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# including X-chrom, get all the gene info from both datasets\n",
    "illuminagenes = pd.read_csv('FinalData/Illumina_SFARI_mutation_matrix.txt', sep=',')\n",
    "CG_genes = pd.read_csv('FinalData/SFARIgene_mutation_matrix.txt', sep=',')\n",
    "CG_genes = CG_genes.drop(columns=['Unnamed: 0'])\n",
    "both_genes = pd.concat([illuminagenes, CG_genes], join = 'inner')\n",
    "both_genes = both_genes.reset_index(drop=True)\n",
    "\n",
    "all_genes_aut = both_genes[both_genes['PHENOTYPE']==2]\n",
    "all_genes_neuro = both_genes[both_genes['PHENOTYPE']==1]\n",
    "all_genes_aut_fem = all_genes_aut[all_genes_aut['SEX']==2]\n",
    "all_genes_aut_mal = all_genes_aut[all_genes_aut['SEX']==1]\n",
    "all_genes_neuro_fem = all_genes_neuro[all_genes_neuro['SEX']==2]\n",
    "all_genes_neuro_mal = all_genes_neuro[all_genes_neuro['SEX']==1]\n",
    "\n",
    "matrices = [all_genes_aut_fem, all_genes_aut_mal, all_genes_neuro_fem, all_genes_neuro_mal]\n",
    "matrices = [df.drop(columns=['PHENOTYPE', 'SEX', 'IID']) for df in matrices]\n",
    "all_genes_aut_fem, all_genes_aut_mal, all_genes_neuro_fem, all_genes_neuro_mal = matrices\n",
    "\n",
    "mal_sig_genes, mal_insig_genes = loeuf_score_calc(all_genes_aut_mal, all_genes_neuro_mal)\n",
    "fem_sig_genes, fem_insig_genes = loeuf_score_calc(all_genes_aut_fem, all_genes_neuro_fem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcdffb08-0179-4b1a-8ccb-37e736c27ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot a violin plot of the distribution of genes that are significant (male or female) and those that aren't\n",
    "def loeuf_plot(signif_genes, insignif_genes, typ):\n",
    "    stat, pval_loeuf = mannwhitneyu(signif_genes['oe_lof_upper'], insignif_genes['oe_lof_upper'], alternative='two-sided')\n",
    "\n",
    "    insignif_genes['group'] = 'Insignificant genes'\n",
    "    signif_genes['group'] = 'Significant genes'\n",
    "\n",
    "    genes_loeuf = pd.concat([insignif_genes, signif_genes])\n",
    "\n",
    "    sns.violinplot(x='group', y='oe_lof_upper', data=genes_loeuf, inner='box')\n",
    "    sns.boxplot(x='group', y='oe_lof_upper', data=genes_loeuf, width=0.1, showcaps=True, \n",
    "            boxprops={'facecolor':'none', 'edgecolor':'black'},\n",
    "            whiskerprops={'color':'black', 'linewidth':2},       \n",
    "            capprops={'color':'black', 'linewidth':2},           \n",
    "            medianprops={'color':'black', 'linewidth':2},       \n",
    "            saturation=0.75)\n",
    "    plt.title('LOEUF scores of SFARI genes' + ' - ' + typ)\n",
    "    plt.xlabel('Genes')\n",
    "    plt.ylabel('LOEUF scores')\n",
    "    plt.show()\n",
    "    return(pval_loeuf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f73a66bf-fc63-4f21-b203-d32640c98b49",
   "metadata": {},
   "outputs": [],
   "source": [
    "loeuf_plot(mal_sig_genes, mal_insig_genes, 'male')\n",
    "loeuf_plot(fem_sig_genes, fem_insig_genes, 'female')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed22eec2-d729-4c46-888a-33e44e23f129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find which genes are both male and female or just one\n",
    "fm_gene_overlap = set(fem_sig_genes.index.values.tolist()) & set(mal_sig_genes.index.values.tolist())\n",
    "fm_only_gene = set(fem_sig_genes.index.values.tolist()) - set(mal_sig_genes.index.values.tolist())\n",
    "ml_only_gene = set(mal_sig_genes.index.values.tolist()) - set(fem_sig_genes.index.values.tolist())\n",
    "both_genes = fem_sig_genes[~fem_sig_genes.index.isin(fm_only_gene)]\n",
    "mal_only_genes = mal_sig_genes.loc[list(ml_only_gene)]\n",
    "fem_only_genes = fem_sig_genes.loc[list(fm_only_gene)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7987ce5-f39c-4a36-b25d-de02012bfea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the significant genes for ABAEnrichment analysis in R\n",
    "save = False\n",
    "if save:\n",
    "    fem_sig_genes.index.to_series().to_csv('ABA_enrichment/female_sig_SFARI_genes.txt', index=False, header=False)\n",
    "    mal_sig_genes.index.to_series().to_csv('ABA_enrichment/male_sig_SFARI_genes.txt', index=False, header=False)\n",
    "    all_sig_genes.index.to_series().to_csv('ABA_enrichment/all_sig_SFARI_genes.txt', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03196b62-61d9-4cae-af49-ae24834973df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get shap scores for genes\n",
    "explainerG = shap.Explainer(tuned_rf)\n",
    "X_test_G= testG.drop(columns=['PHENOTYPE'])\n",
    "shap_valuesG = explainerG.shap_values(X_test_G)\n",
    "shap_class_valuesG = shap_valuesG[:, :, 1]\n",
    "mean_shap_feature_valuesG = pd.DataFrame(shap_class_valuesG, columns=X_test_G.columns)\n",
    "mean_shap_feature_valuesG = mean_shap_feature_valuesG.abs().mean(axis=0).sort_values(ascending=False)\n",
    "mean_shap_feature_valuesG = mean_shap_feature_valuesG.reset_index()\n",
    "mean_shap_feature_valuesG.columns = ['gene', 'SHAP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5363730-73e3-40b2-8a74-11c1d19c81ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare males and females gene shap values\n",
    "pos_classG = shap_valuesG[:,:,1]\n",
    "shapG = pd.DataFrame(pos_classG, columns=X_test_G.columns)\n",
    "shapG['sex'] = testP['SEX'].values\n",
    "\n",
    "shap_femaleG = shapG[shapG['sex'] == 2].drop('sex', axis=1)\n",
    "shap_maleG = shapG[shapG['sex'] == 1].drop('sex', axis=1)  \n",
    "mean_shap_femaleG = shap_femaleG.abs().mean(axis=0).sort_values(ascending=False)\n",
    "mean_shap_femaleG = mean_shap_femaleG.reset_index()\n",
    "mean_shap_femaleG.columns = ['gene', 'SHAP']\n",
    "\n",
    "mean_shap_maleG = shap_maleG.abs().mean(axis=0).sort_values(ascending=False)\n",
    "mean_shap_maleG = mean_shap_maleG.reset_index()\n",
    "mean_shap_maleG.columns = ['gene', 'SHAP']\n",
    "\n",
    "#plot the top 5 contributing genes\n",
    "mean_shap_femaleGenes = shap_femaleG.abs().mean()\n",
    "mean_shap_maleGenes = shap_maleG.abs().mean()\n",
    "\n",
    "shap_comparisonG = pd.DataFrame({\n",
    "    'female': mean_shap_femaleGenes,\n",
    "    'male': mean_shap_maleGenes\n",
    "})\n",
    "\n",
    "top_n = 5\n",
    "shap_comparisonG.sort_values(by='male', ascending=False).head(top_n).plot(kind='bar', figsize=(10, 6))\n",
    "plt.title(f'Top {top_n} Contributing Genes (Female vs Male)')\n",
    "plt.ylabel('Mean abs(SHAP Value)')\n",
    "save = False\n",
    "if save:\n",
    "    plt.savefig('top_genes_balanced.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6b2799a-6e79-4322-bfb0-14811c61b3a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python pycaret env",
   "language": "python",
   "name": "your_env_name"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
